core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,185 @@
+                })
+                // We only decode function calls that have an array aggregation as only parameter (because decoding is expensive)
+                .filter(entry -> {
+                    FunctionCall functionCall = (FunctionCall) entry.getValue();
+                    ResolvedFunction resolvedFunction = metadata.decodeFunction(functionCall.getName());
+                    return resolvedFunction.getSignature().getName().equals(ArrayDistinctFunction.NAME);
+                })
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        Set<Symbol> referencedOutsideArrayDistinctAggregateSymbols = projectNode.getAssignments().entrySet().stream()
+                .filter(entry -> !arrayDistinctCalls.containsKey(entry.getKey()))
+                .flatMap(entry -> SymbolsExtractor.extractUnique(entry.getValue()).stream())
+                .filter(arrayAggSymbols::contains)
+                .collect(toImmutableSet());
+
+        Map<Symbol, Aggregation> arrayAggregationsWithArrayDistinct = aggregationNode.getAggregations().entrySet().stream()
+                .filter(entry -> !referencedOutsideArrayDistinctAggregateSymbols.contains(entry.getKey()))
+                .filter(entry -> arrayDistinctCalls.values().stream().anyMatch(arrayDistinctCall -> {
+                    Symbol symbol = Symbol.from(getOnlyElement(((FunctionCall) arrayDistinctCall).getArguments()));
+                    return symbol.equals(entry.getKey());
+                }))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        for (Map.Entry<Symbol, Aggregation> entry : arrayAggregationsWithArrayDistinct.entrySet()) {
+            Symbol outputSymbol = entry.getKey();
+            Aggregation aggregation = entry.getValue();
+
+            if (!aggregation.isDistinct()) {
+                matchedAggregations.add(outputSymbol);
+                Aggregation newAggregation = new Aggregation(
+                        aggregation.getResolvedFunction(),
+                        aggregation.getArguments(),
+                        true,
+                        aggregation.getFilter(),
+                        aggregation.getOrderingScheme(),
+                        aggregation.getMask());
+                aggregationsBuilder.put(outputSymbol, newAggregation);
+            }
+
+            Set<Symbol> arrayDistinctSymbols = arrayDistinctCalls.entrySet().stream()
+                    .filter(arrayDistinctCall -> {
+                        FunctionCall functionCall = (FunctionCall) arrayDistinctCall.getValue();
+                        Expression functionArgument = getOnlyElement(functionCall.getArguments());
+                        return Symbol.from(functionArgument).equals(outputSymbol);
+                    })
+                    .map(Map.Entry::getKey)
+                    .collect(toImmutableSet());
```

Comment:
Instead of looping through the assignments for each aggregation, you could run through them once and retrieve those that refer to `arrayAggregationsToMakeDistinct`.